### PR TITLE
store history revisions for sidecarset

### LINKED ISSUE #710: [feature request]Support version control for SidecarSet
**What would you like to be added**:
Support version control for SidecarSet ( e.g., using ControllerRevision).

**Why is this needed**:

- Perform a rollback on a SidecarSet;
- Specify version  when SidecarSet injection;

### PR DESCRIPTION
**Pull request by veophi**
Signed-off-by: veophi <vec.g.sun@gmail.com>

### Ⅰ. Describe what this PR does
1. Store history revisions for sidecarset.
- Store sidecarset.spec using controllerRevision;
- All controllerRevisions of sidecarset will be stored in the namespace of `kruise-manager`;
- We set a limited number of stored revisions (default is 10).

2. Add revision annotation for matched pods. 
- This annotation can remind users which sidecar is not the latest; 
- This annotation may be used in the future(e.g., for the future plan of replacing update operation with patch).

### Ⅱ. Does this pull request fix one issue?

fixes #710

### Ⅲ. What does it look like?
image

### CODE REVIEW

File: pkg/controller/sidecarset/sidecarset_processor.go [hunk 1]
```diff
@@ -284,6 +330,99 @@ func (p *Processor) getSelectedPods(namespaces []string, selector labels.Selecto
...
+
+	return latestRevision, collisionCount, nil
+}
+
+func (p *Processor) truncateHistory(revisions []*apps.ControllerRevision) error {
+	limited := integer.IntMax(0, maxStoredRevisions)
+	if len(revisions) == 0 || len(revisions) <= limited {
+		return nil
+	}
+	if err := p.historyController.DeleteControllerRevision(revisions[0]); err != nil {
```
Comment (FillZpp) [inline pkg/controller/sidecarset/sidecarset_processor.go:415]: Should not delete the ControllerRevisions that have existing pods in these revisions.

File: pkg/controller/sidecarset/sidecarset_processor.go [hunk 2]
```diff
@@ -284,6 +330,99 @@ func (p *Processor) getSelectedPods(namespaces []string, selector labels.Selecto
...
+		return err
+	}
+	return nil
+}
+
+func getCurrentStatus(control sidecarcontrol.SidecarControl, pods []*corev1.Pod,
```
Comment (FillZpp) [inline pkg/controller/sidecarset/sidecarset_processor.go:441]: Just add `latestRevision` and `collisionCount` into `calculateStatus`. No need to define getCurrentStatus and setStatusRevisionInfo

File: pkg/controller/sidecarset/sidecarset_processor.go [hunk 3]
```diff
@@ -189,11 +200,46 @@ func (p *Processor) updatePodSidecarAndHash(control sidecarcontrol.SidecarContro
...
+	// older sidecarSets don't have controllerRevision
+	sidecarSetNames, nameOk := pod.Annotations[sidecarcontrol.SidecarSetListAnnotation]
+	sidecarCount := len(strings.Split(sidecarSetNames, ","))
+	if !nameOk || sidecarCount == 0 || len(sidecarSetRevisions) != sidecarCount {
+		sidecarSets := p.listMatchedSidecarSets(pod)
+		for _, sidecar := range sidecarSets {
+			// in the future, we will support revision control for sidecar injection,
+			// some pods may be injected old sidecarSet
+			if _, ok := sidecarSetRevisions[sidecar]; !ok {
+				sidecarSetRevisions[sidecar] = sidecarcontrol.SidecarLatestRevision
```
Comment (FillZpp) [inline pkg/controller/sidecarset/sidecarset_processor.go:223]: Don't really understand what does the `SidecarLatestRevision` for?

File: pkg/controller/sidecarset/sidecarset_processor.go [hunk 4]
```diff
@@ -189,11 +200,46 @@ func (p *Processor) updatePodSidecarAndHash(control sidecarcontrol.SidecarContro
...
+		}
+	}
+
+	// older pods don't have SidecarSetListAnnotation
+	// older sidecarSets don't have controllerRevision
+	sidecarSetNames, nameOk := pod.Annotations[sidecarcontrol.SidecarSetListAnnotation]
+	sidecarCount := len(strings.Split(sidecarSetNames, ","))
+	if !nameOk || sidecarCount == 0 || len(sidecarSetRevisions) != sidecarCount {
+		sidecarSets := p.listMatchedSidecarSets(pod)
+		for _, sidecar := range sidecarSets {
```
Comment (FillZpp) [inline pkg/controller/sidecarset/sidecarset_processor.go:219]: Maybe here should only manage the sidecar revision of this sidecarset in the pod annotation.

File: apis/apps/v1alpha1/sidecarset_types.go [hunk 5]
```diff
@@ -189,6 +189,17 @@ type SidecarSetStatus struct {
...
+	// Revision, if not empty, indicates the latest revision name of the SidecarSet.
+	// If a Pod was injected this latest revision of sidecarSet,
+	// Pod.Annotations["kruise.io/sidecarset-injected-revision"] will be
+	// '{"<sidecar-name>": "latest"}',  else it will be a specific revision name
+	LatestRevision string `json:"updateRevision,omitempty"`
+
+	// CollisionCount is the count of hash collisions for the SidecarSet. The SidecarSet controller
+	// uses this field as a collision avoidance mechanism when it needs to create the name for the
+	// newest ControllerRevision.
+	CollisionCount *int32 `json:"collisionCount,omitempty"`
```
[the remark this answers is not in the document]
Comment (veophi) [reply apis/apps/v1alpha1/sidecarset_types.go:202]: sounds good

File: pkg/controller/sidecarset/sidecarset_processor.go (hunk 4 above)
Comment (veophi) [reply pkg/controller/sidecarset/sidecarset_processor.go:219]: I continue to use the name list annotation just for compatibility.

File: pkg/controller/sidecarset/sidecarset_processor.go (hunk 3 above)
Comment (veophi) [reply pkg/controller/sidecarset/sidecarset_processor.go:223]: 1. Some older pods may not have corresponding sidecar history revision, and they must the latest, so we use this to mark it.
2. If users want to check whether the latest sidecar has been injected, they do not need to check the historical revision number.

File: pkg/controller/sidecarset/sidecarset_processor.go (hunk 2 above)
Comment (veophi) [reply pkg/controller/sidecarset/sidecarset_processor.go:441]: ok

File: pkg/controller/sidecarset/sidecarset_processor.go [hunk 6]
```diff
@@ -284,6 +330,99 @@ func (p *Processor) getSelectedPods(namespaces []string, selector labels.Selecto
...
+	)
+	if err != nil {
+		klog.Errorf("Failed to convert labels to selector, err %v, name %v", err, s.Name)
+		return nil, collisionCount, err
+	}
+
+	sidecarSetClone := s.DeepCopy()
+	// set controllerRevision namespace, this namespace must have very
+	// strict permissions for average users. Here use namespace of kruise-manager
+	sidecarSetClone.SetNamespace(webhookutil.GetNamespace())
```
[the remark this answers is not in the document]
Comment (veophi) [reply pkg/controller/sidecarset/sidecarset_processor.go:349]: in `kubernetes/pkg/controller/history/controller_history.go`, `func  CreateControllerRevision` use parent's namespace to create ControllerRevision, so we have to set sidecarset namespace here.

[the remark this answers is not in the document]
Comment (veophi) [reply pkg/controller/sidecarset/sidecarset_processor.go:349]: Or we might copy these codes and modify them.

File: pkg/controller/sidecarset/sidecarset_processor.go (hunk 1 above)
Comment (veophi) [reply pkg/controller/sidecarset/sidecarset_processor.go:415]: So should we count references for each revisions?

File: pkg/control/sidecarcontrol/history_control.go [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+		revision,
+		collisionCount)
+	if err != nil {
+		return nil, err
+	}
+	cr.SetNamespace(s.Namespace)
+	if cr.Labels == nil {
+		cr.Labels = make(map[string]string)
+	}
+	cr.SetLabels(r.GetRevisionLabelSelector(s).MatchLabels)
```
[the remark this answers is not in the document]
Comment (veophi) [reply pkg/control/sidecarcontrol/history_control.go:68]: ok

File: pkg/control/sidecarcontrol/history_control.go [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+// recorded patches.
+func (r *realControl) getPatch(s *appsv1alpha1.SidecarSet) ([]byte, error) {
+	str, err := runtime.Encode(patchCodec, s)
+	if err != nil {
+		return nil, err
+	}
+	var raw map[string]interface{}
+	_ = json.Unmarshal(str, &raw)
+	objCopy := make(map[string]interface{})
+	objCopy["spec"] = raw["spec"].(map[string]interface{})
```
[the remark this answers is not in the document]
Comment (veophi) [reply pkg/control/sidecarcontrol/history_control.go:90]: ok

File: pkg/control/sidecarcontrol/util.go [hunk 9]
```diff
@@ -38,6 +38,10 @@ const (
 	// SidecarSetHashWithoutImageAnnotation represents the key of a sidecarset hash without images of sidecar
 	SidecarSetHashWithoutImageAnnotation = "kruise.io/sidecarset-hash-without-image"
 
+	// SidecarSetHashAnnotation represents the revision of a sidecarSet
+	SidecarSetRevisionAnnotation = "kruise.io/sidecarset-injected-revision"
```
Comment (zmberg) [inline pkg/control/sidecarcontrol/util.go:42]: Why not put the revision in SidecarSetUpgradeSpec?

File: pkg/controller/sidecarset/sidecarset_processor.go [hunk 10]
```diff
@@ -204,7 +250,7 @@ func (p *Processor) listMatchedSidecarSets(pod *corev1.Pod) string {
 		}
 	}
 
-	return strings.Join(sidecarSetNames, ",")
+	return sidecarSetNames
 }
 
 func (p *Processor) updateSidecarSetStatus(sidecarSet *appsv1alpha1.SidecarSet, status *appsv1alpha1.SidecarSetStatus) error {
```
Comment (zmberg) [inline pkg/controller/sidecarset/sidecarset_processor.go:256]: Is function inconsistentStatus necessary to determine revision filed?

File: pkg/controller/sidecarset/sidecarset_processor.go (hunk 3 above)
Comment (veophi) [reply pkg/controller/sidecarset/sidecarset_processor.go:223]: I realize I'm wrong and I'll fix it as soon as possible.

File: pkg/controller/sidecarset/sidecarset_processor.go [hunk 11]
```diff
@@ -127,13 +142,13 @@ func (p *Processor) UpdateSidecarSet(sidecarSet *appsv1alpha1.SidecarSet) (recon
 	}
 
 	// 7. upgrade pod sidecar
-	if err := p.updatePods(control, pods); err != nil {
+	if err := p.updatePods(control, pods, latestRevision); err != nil {
```
Comment (zmberg) [inline pkg/controller/sidecarset/sidecarset_processor.go:145]: sidecarSet.status already contains the latestRevison, so don't need to add additional parameters latestRevision

File: pkg/controller/sidecarset/sidecarset_processor.go [hunk 12]
```diff
@@ -157,19 +172,15 @@ func (p *Processor) updatePods(control sidecarcontrol.SidecarControl, pods []*co
...
-		// which is to improve the performance of the sidecarSet controller
-		sidecarSetNames, ok := podClone.Annotations[sidecarcontrol.SidecarSetListAnnotation]
-		if !ok || len(sidecarSetNames) == 0 {
-			podClone.Annotations[sidecarcontrol.SidecarSetListAnnotation] = p.listMatchedSidecarSets(podClone)
-		}
-
+		// update SidecarSetRevisionAnnotation and SidecarSetRevisionAnnotation of pod
+		// These annotations improve the performance of the sidecarSet controller
+		// In the future, SidecarSetRevisionAnnotation may help sidecarSet controller replace update operation with patch
+		p.updatePodAnnotationsAboutSidecarNameAndRevision(control, podClone, latestRevision)
```
Comment (zmberg) [inline pkg/controller/sidecarset/sidecarset_processor.go:183]: I think you can update revision in function control.updatePodSidecarSetHash

File: pkg/controller/sidecarset/sidecarset_processor.go (hunk 11 above)
Comment (veophi) [reply pkg/controller/sidecarset/sidecarset_processor.go:145]: Do you mean we should execute "get status" in this function?

**Pull request merged**